op-e2e/migration_test.go

```diff
@@ -0,0 +1,280 @@
+	"math/big"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth/ethconfig"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/ethclient/gethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/kroma-network/zktrie/trie"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ethereum-optimism/optimism/op-e2e/e2eutils/geth"
+	rollupNode "github.com/ethereum-optimism/optimism/op-node/node"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/driver"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/sync"
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+	"github.com/kroma-network/kroma/kroma-bindings/predeploys"
+	"github.com/kroma-network/kroma/op-e2e/e2eutils/wait"
+)
+
+func TestMigration(t *testing.T) {
+	InitParallel(t)
+
+	zero := hexutil.Uint64(0)
+	one := hexutil.Uint64(1)
+	mptTimeOffset := hexutil.Uint64(60)
+
+	cfg := DefaultSystemConfig(t)
+	cfg.DeployConfig.L2GenesisDeltaTimeOffset = &zero
+	cfg.DeployConfig.L2GenesisEcotoneTimeOffset = &one
+	cfg.DeployConfig.L2GenesisKromaMPTTimeOffset = &mptTimeOffset
+
+	// Setup historical rpc node.
+	historicalRpcPort := 8045
+	cfg.Nodes["historical"] = &rollupNode.Config{
+		Driver: driver.Config{
+			VerifierConfDepth:  0,
+			SequencerConfDepth: 0,
+			SequencerEnabled:   false,
+		},
+		L1EpochPollInterval:         time.Second * 4,
+		RuntimeConfigReloadInterval: time.Minute * 10,
+		ConfigPersistence:           &rollupNode.DisabledConfigPersistence{},
+		Sync:                        sync.Config{SyncMode: sync.CLSync},
+	}
+	cfg.Loggers["historical"] = testlog.Logger(t, log.LevelInfo).New("role", "historical")
+	cfg.GethOptions["historical"] = append(cfg.GethOptions["historical"], []geth.GethOption{
+		func(ethCfg *ethconfig.Config, nodeCfg *node.Config) error {
+			nodeCfg.HTTPPort = historicalRpcPort
+			nodeCfg.HTTPModules = []string{"debug", "eth"}
+			nodeCfg.HTTPHost = "127.0.0.1"
+			return nil
+		},
+	}...)
+
+	// Set historical rpc endpoint.
+	for name := range cfg.Nodes {
+		name := name
+		cfg.GethOptions[name] = append(cfg.GethOptions[name], []geth.GethOption{
+			func(ethCfg *ethconfig.Config, nodeCfg *node.Config) error {
+				// Since the migration process requires preimages, enable storing preimage option.
+				ethCfg.Preimages = true
+				ethCfg.RollupHistoricalRPC = fmt.Sprintf("http://127.0.0.1:%d", historicalRpcPort)
+				if name == "historical" {
+					ethCfg.RollupHistoricalRPC = ""
+					ethCfg.DisableMPTMigration = true
+				}
+				// Deep copy the genesis
+				dst := &core.Genesis{}
+				b, _ := json.Marshal(ethCfg.Genesis)
+				err := json.Unmarshal(b, dst)
+				if err != nil {
+					return err
+				}
+				ethCfg.Genesis = dst
```

Comment:
In e2e, when there are multiple nodes, if one node finalizes transition to mpt first, the shared genesis causes other nodes to also have their genesis changed. This is to avoid such an issue.